### PR TITLE
Bump to `typescript@5.1`

### PR DESCRIPTION
**Pull request by styfle**
We need to upgrade TypeScript to land the following fixes:

- https://github.com/vercel/next.js/pull/50289
- https://github.com/vercel/next.js/pull/48018

And it should fix async components:
- https://twitter.com/sebsilbermann/status/1664356039876124672

Also see this related TS 5.0 issue:

- https://github.com/microsoft/TypeScript/issues/53402

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 0dcf74d4f3808fe2a8f02638615b7379b377ea15

`pnpm testheadless test/integration/typescript-version-warning/test/index.test.js` 
- Minimum TypeScript Warning > should show warning during next build with old version
- Minimum TypeScript Warning > should show warning during next dev with old version

<details>
<summary>Expand output</summary>

  ● Minimum TypeScript Warning › should show warning during next build with old version

    expect(received).toContain(expected) // indexOf

    Expected substring: "Minimum recommended TypeScript version is"
    Received string:    "- info Linting and checking validity of types...
    - info Creating an optimized production build...
    - info Using locally built binary of @next/swc
    - info Compiled successfully
    - info Collecting page data...
    - info Generating static pages (0/3)
    - info Generating static pages (3/3)
    - info Finalizing page optimization...·
    Route (pages)                              Size     First Load JS
    ┌ ○ /                                      261 B          73.2 kB
    └ ○ /404                                   181 B          73.2 kB
    + First Load JS shared by all              73 kB
      ├ chunks/framework-e25e86378fd2828c.js   45.2 kB
      ├ chunks/main-582416ee48ed17ce.js        26.8 kB
      ├ chunks/pages/_app-891afadd7b943fb0.js  195 B
      └ chunks/webpack-4e7214a60fad8e88.js     712 B·
    ○  (Static)  automatically rendered as static HTML (uses no initial props)·
    - warn The Next.js plugin was not detected in your ESLint configuration. See https://nextjs.org/docs/basic-features/eslint#migrating-existing-config
    DeprecationWarning: 'originalKeywordKind' has been deprecated since v5.0.0 and will no longer be usable after v5.2.0. Use 'identifierToKeywordKind(identifier)' instead.
    "

      13 |       stdout: true,
      14 |     })
    > 15 |     expect(res.stdout + res.stderr).toContain(
         |                                     ^
      16 |       'Minimum recommended TypeScript version is'
      17 |     )
      18 |   })

      at Object.<anonymous> (integration/typescript-version-warning/test/index.test.js:15:37)

  ● Minimum TypeScript Warning › should show warning during next dev with old version

    expect(received).toContain(expected) // indexOf

    Expected substring: "Minimum recommended TypeScript version is"
    Received string:    "- ready started server on 0.0.0.0:43365, url: http://localhost:43365
    - info Using locally built binary of @next/swc
    - event compiled client and server successfully in 210 ms (17 modules)
    "

      30 |     await killApp(app)
      31 |
    > 32 |     expect(output).toContain('Minimum recommended TypeScript version is')
         |                    ^
      33 |   })
      34 |
      35 |   it('should not show warning during next build with new version', async () => {

      at Object.<anonymous> (integration/typescript-version-warning/test/index.test.js:32:20)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: test/e2e/typescript-version-warning/typescript-version-warning.test.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        expect(1).toBe(1)
+      })
+    } else {
+      it('should print warning when old typescript version is used with next build', async () => {
+        await next.start().catch(() => {})
+        expect(next.cliOutput).toContain(
+          'warn Minimum recommended TypeScript version is v4.5.2, older versions can potentially be incompatible with Next.js. Detected: 4.0.6'
+        )
+      })
+    }
```
Comment (ForsakenHarmony) [inline test/e2e/typescript-version-warning/typescript-version-warning.test.ts:25]:
```suggestion
    if (isNextDeploy || isNextDev) {
      it('should skip', () => {})
      return
    }

    it('should print warning when old typescript version is used with next build', async () => {
      await next.start().catch(() => {})
      expect(next.cliOutput).toContain(
        'warn Minimum recommended TypeScript version is v4.5.2, older versions can potentially be incompatible with Next.js. Detected: 4.0.6'
      )
    })
```

File: test/e2e/typescript-version-no-warning/typescript-version-no-warning.test.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+createNextDescribe(
+  'typescript-version-no-warning',
+  {
+    files: __dirname,
+    skipStart: true,
+    skipDeployment: true,
+  },
+  ({ next, isNextDeploy, isNextDev }) => {
+    if (isNextDeploy || isNextDev) {
```
Comment (ForsakenHarmony) [inline test/e2e/typescript-version-no-warning/typescript-version-no-warning.test.ts:11]: same as below(?)

**Review by ForsakenHarmony — DISMISSED** [review]

File: test/e2e/typescript-version-warning/typescript-version-warning.test.ts (hunk 1 above)
Comment (styfle) [reply test/e2e/typescript-version-warning/typescript-version-warning.test.ts]: Fixed in [81c0f18](https://github.com/vercel/next.js/pull/50557/commits/81c0f18964d450b1df7576d69d9a2931307e0de7)

**Review by ForsakenHarmony — APPROVED** [review]

**Pull request merged**